pkg/executor/executor.go

```diff
@@ -410,19 +411,57 @@ type DDLJobRetriever struct {
 	activeRoles    []*auth.RoleIdentity
 	cacheJobs      []*model.Job
 	TZLoc          *time.Location
+	extractor      base.MemTablePredicateExtractor
 }
 
 func (e *DDLJobRetriever) initial(txn kv.Transaction, sess sessionctx.Context) error {
-	m := meta.NewMeta(txn)
-	jobs, err := ddl.GetAllDDLJobs(sess)
-	if err != nil {
-		return err
+	skipRunningJobs := false
+	skipHistoryJobs := false
+	schemaNames := set.NewStringSet()
+	tableNames := set.NewStringSet()
+
+	ex, ok := e.extractor.(*plannercore.InfoSchemaDDLExtractor)
+	if ok {
+		// Using state to determine whether we can skip checking running/history ddl jobs
+		if states, ok := ex.ColPredicates["state"]; ok {
+			skipHistoryJobs = true
+			skipRunningJobs = true
+			states.IterateWith(func(s string) {
+				ss := strings.ToLower(s)
+				if ss == "cancelled" || ss == "synced" {
+					skipHistoryJobs = false
+				} else {
+					skipRunningJobs = false
+				}
+			})
+		}
+
+		schemaNames = ex.ColPredicates["db_name"]
+		tableNames = ex.ColPredicates["table_name"]
 	}
-	e.historyJobIter, err = ddl.GetLastHistoryDDLJobsIterator(m)
-	if err != nil {
-		return err
+
+	var err error
+
+	if !skipRunningJobs {
+		// We cannot use table_id and schema_id to construct predicates for the tidb_ddl_job table.
+		// For instance, in the case of the SQL like `create table t(id int)`,
+		// the tableInfo for 't' will not be available in the infoschema until the job is completed.
+		// As a result, we cannot retrieve its table_id.
+		e.runningJobs, err = ddl.GetAllDDLJobs(sess)
```

Comment:
we can think about how to make this part for streaming, like not draining the result at first, cause we don't know whether there is jobNum limit outside as we say.

the below historyJobIter is a quite good example